fix(loader): randomized AppImage path pollutes luac cache #31165

shmerl opened the pull request:
Fixes infinite growth of number of *.luac files in the cache and actual cache file misses caused by randomness of filenames when using AppImage with loader enabled.

fix #31165

shmerl committed:
fix(loader): prevent randomness in AppImage luac files in the cache #31165

shmerl commented:
Hm, it fails some checks like this:

```
runtime/lua/vim/loader.lua:122:11 [Warning] Can not infer type. (no-unknown)
        local idx = name:find('/usr/', 1, true)
```

I'm not exactly sure what it doesn't like. Any suggestion how to prevent that? The `idx` should be a number of `nil`.

justinmk commented on runtime/lua/vim/loader.lua:
```diff
@@ -116,6 +116,13 @@ end
+  -- Simplify the name of AppImage mounted file, to prevent
+  -- random part in it from creating cache misses.
+  if os.getenv('APPIMAGE') ~= nil then
+    local idx = name:find('/usr/', 1, true) --- @type integer?
```
look at examples in the codebase, if in doubt
```suggestion
    local idx = name:find('/usr/', 1, true) ---@type integer?
```

justinmk commented on runtime/lua/vim/loader.lua:
```diff
@@ -116,6 +116,13 @@ end
+  -- Simplify the name of AppImage mounted file, to prevent
+  -- random part in it from creating cache misses.
+  if os.getenv('APPIMAGE') ~= nil then
+    local idx = name:find('/usr/', 1, true) --- @type integer?
+    name = idx and name:sub(idx) or name
```
1. this can be simplified using match()
2. always show an input/output example, it communicates so much in a very concise way.

```suggestion
    -- Avoid cache pollution caused by AppImage randomizing the program root. #31165
    -- "/tmp/.mount_nvimAmpHPH/usr/share/nvim/runtime" => "/usr/share/nvim/runtime"
    name = name:match("(/usr/.*)") or name
```

shmerl commented:
> I pushed a change, can you confirm it works so we can wrap this up

Never did it before, how can I build an appimage from my branch?

shmerl commented:
OK, I managed to repack things using appimage-tool using the patched file. Checking.

shmerl commented:
Weird, even with fixed file I still see this accumulation:

```
Only in luac: %2ftmp%2f.mount_NeovimHiB7yR%2fusr%2fshare%2fnvim%2fruntime%2ffiletype.luac
Only in luac: %2ftmp%2f.mount_NeovimHiB7yR%2fusr%2fshare%2fnvim%2fruntime%2flua%2fvim%2ftermcap.luac
Only in luac: %2ftmp%2f.mount_NeovimHiB7yR%2fusr%2fshare%2fnvim%2fruntime%2fplugin%2feditorconfig.luac
Only in luac: %2ftmp%2f.mount_NeovimHiB7yR%2fusr%2fshare%2fnvim%2fruntime%2fplugin%2fman.luac
Only in luac: %2ftmp%2f.mount_NeovimHiB7yR%2fusr%2fshare%2fnvim%2fruntime%2fplugin%2fosc52.luac
Only in luac: %2ftmp%2f.mount_NeovimHiB7yR%2fusr%2fshare%2fnvim%2fruntime%2fplugin%2ftohtml.luac
```

FYI, the way I tested it. Extracted existing appimage release:

```
nvim --appimage-extract
```

Replaced `loader.lua` and then repacked things back with

```
wcurl https://github.com/AppImage/AppImageKit/releases/download/continuous/appimagetool-x86_64.AppImage

./appimagetool-x86_64.AppImage --no-appstream squashfs-root 
```

Really not sure why it's not helping though.

justinmk commented:
Try removing the `os.getenv('APPIMAGE')` check

shmerl commented:
Nope, that's not it. It's still using that cursed path. I think something else is happening besides loader may be? Also, I can confirm that the variable is set if I read this from inside neovim itself

```
= os.getenv('APPIMAGE'
```

Returns binary path.

justinmk commented:
To double check that the appimage "repack" is actually working, can you try throwing an error in the loader somewhere and confirm that the error is thrown after repack :)

justinmk commented:
Just a guess, but confirm that `enable=true` here https://github.com/neovim/neovim/blob/2f78ff816b03661b5f74d0624e973eaca0d64ef1/runtime/lua/vim/loader.lua#L424

else you are hitting the "standard" path in that if-else, which does not use the custom loader.

shmerl commented:
I added something like this:

```
function M.enable(enable)
  error("enable = " .. enable)
...
```

But I don't see any exception thrown. I don't get it, is this module even used at all?

shmerl commented:
> loader.lua may be preload as bytecode

Where would it come from though? I don't see any luac files in the appimage tree.

shmerl commented:
I was actually starting re-packed appimage since that should trigger the env variable check. I can try using the above method.

shmerl commented:
@justinmk: Just got to testing it using your method with providing the modified runtime with the fix. It works!

I.e. I unpacked the appimage to `$HOME/tmp/neovim_debug`, applied the change to loader.lua and tried this:
```
VIMRUNTIME=$HOME/tmp/neovim_debug/squashfs-root/usr/share/nvim/runtime nvim --luamod-dev
```

Now those previously problematic files are created without the bad prefix like this:

```
%2fusr%2fshare%2fnvim%2fruntime%2ffiletype.luac
```

So I can confirm that the fix helps.

justinmk commented:
> That's a bit over my head now,

try an AI to get something started. You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#filter-by-file) like this:

    TEST_FILE=test/functional/lua/loader_spec.lua make functionaltest


> I'll have to take a closer look how to make a new test. Do you suggest to add it to some existing unit test file or make a new one?

`test/functional/lua/loader_spec.lua`

shmerl commented:
I'll look into the test a bit later.
